server/src/test/java/org/elasticsearch/index/mapper/MapperServiceTests.java

```diff
@@ -1467,4 +1455,262 @@ public void testMergeUntilLimitCapacityOnlyForParent() throws IOException {
         assertNull(mapper.mappers().getMapper("parent.child"));
     }
 
+    public void testAutoFlattenObjectsSubobjectsTopLevelMerge() throws IOException {
+        CompressedXContent mapping1 = new CompressedXContent("""
+            {
+              "subobjects": false
+            }""");
+
+        CompressedXContent mapping2 = new CompressedXContent("""
+            {
+              "properties": {
+                "parent": {
+                  "properties": {
+                    "child": {
+                    "dynamic": true,
+                      "properties": {
+                        "grandchild": {
+                          "type": "keyword"
+                        }
+                      }
+                    }
+                  }
+                }
+              }
+            }""");
+
+        assertMergeEquals(List.of(mapping1, mapping2), """
+            {
+              "_doc" : {
+                "subobjects" : false,
+                "properties" : {
+                  "parent.child.grandchild" : {
+                    "type" : "keyword"
+                  }
+                }
+              }
+            }""");
+    }
+
+    public void testAutoFlattenObjectsSubobjectsMerge() throws IOException {
+        CompressedXContent mapping1 = new CompressedXContent("""
+            {
+              "properties" : {
+                "parent" : {
+                  "properties" : {
+                    "child" : {
+                      "type": "object"
+                    }
+                  }
+                }
+              }
+            }""");
+
+        CompressedXContent mapping2 = new CompressedXContent("""
+            {
+              "properties" : {
+                "parent" : {
+                  "subobjects" : false,
+                  "properties" : {
+                    "child" : {
+                      "properties" : {
+                        "grandchild" : {
+                          "type" : "keyword"
+                        }
+                      }
+                    }
+                  }
+                }
+              }
+            }""");
```

Comment:
this is  good example. And what would have happened if this second mapping only disabled subobjects? would we have auto-flattened the existing child object?

Reply:
> what would have happened if this second mapping only disabled subobjects?

I don't understand. It does disable subobjects:

```json
                "parent" : {
                  "subobjects" : false,
```

Reply:
and the sub-object defined in that path (from the other mapping definition) gets removed?

Reply:
Yes, it get's removed.